lcls_tools/common/data_analysis/bmad_modeling.py

```diff
@@ -0,0 +1,573 @@
+            elif self.data_source == "ARCHIVE":
+                measured = lcls_archiver_restore(pvs, self.date_time)
+            for ii, pv in enumerate(pvs):
+                cmd.append(
+                    f"set dat {datum}[{ii+1}]|meas = {measured[pvs[ii]]}")
+        tao.cmds(cmd)
+        self.tc(tao, f"show dat {datum}")
+
+    def match_twiss(self, tao, variable, datum):
+        match_cmds = ["veto var *", "veto dat *@*",
+                      "set global n_opti_cycles = 912"]
+        match_cmds.append(f"use var {variable}[1:4]")
+        match_cmds.append(f"use dat {datum}[1:4]")
+        tao.cmds(match_cmds)
+        tao.cmd("run")
+        element = self.get_element(tao, datum)
+        self.show_twiss(tao, element, [])
+
+    def get_twiss(self, tao, element, which="design"):
+        """Gets twiss for element, which can be model, desing or base"""
+        result = tao.ele_twiss(element, which=which)
+        return [result[p] for p in ["beta_a", "alpha_a", "beta_b", "alpha_b"]]
+
+    def bmag(self, twiss, twiss_reference):
+        """Calculates BMAG from imput twiss and reference twiss"""
+        beta_a, alpha_a, beta_b, alpha_b = twiss
+        beta_a_ref, alpha_a_ref, beta_b_ref, alpha_b_ref = twiss_reference
+        bmag_a = self.bmag_func(beta_a, alpha_a, beta_a_ref, alpha_a_ref)
+        bmag_b = self.bmag_func(beta_b, alpha_b, beta_b_ref, alpha_b_ref)
+        return (bmag_a, bmag_b)
+
+    def bmag_func(self, bb, ab, bl, al):
+        return 1 / 2 * (bl / bb + bb / bl + bb * bl * (ab / bb - al / bl) ** 2)
+
+    def show_twiss(self, tao, element, datum=[]):
+        # parameters = ["beta_a", "alpha_a", "beta_b", "alpha_b"]
+        if not datum == []:
+            twiss_datum = tao.data_parameter(datum,
+                                             "meas_value")[0].split(";")[1:]
+            twiss_datum = [float(val) for val in twiss_datum]
+        twiss_model = self.get_twiss(tao, element, which="model")
+        twiss_design = self.get_twiss(tao, element)
+        bmag_a, bmag_b = self.bmag(twiss_model, twiss_design)
+        print(f"\n{element} BMAG_X {bmag_a:3.2f}, BMAG_Y {bmag_b:3.2f} ")
+        print(f'{" " * 12} Beta     Alpha   Beta   Alpha ')
+        print(f'{" " * 12}  X       X       Y       Y')
+        print(f'Desing:{" "} ', end="")
+        for val in twiss_design:
+            print(f"{val:8.2f}", end="")
+        print("")
+        print(f'Model:{" " *3}', end="")
+        for val in twiss_model:
+            print(f"{val:8.2f}", end="")
+        print("")
+        if not datum == []:
+            print(f'Measured:{" " * 0}', end="")
+            for val in twiss_datum:
+                print(f"{float(val):8.2f}", end="")
+            bmag_a, bmag_b = self.bmag(twiss_datum, twiss_model)
+            print("\nMeasured to Model:")
+            print(f"\n{element} BMAG_X {bmag_a:3.2f}, BMAG_Y {bmag_b:3.2f}")
+        print("")
+
+    def quad_table(self, tao, pct_lim=1, show_energy=False):
```

Comment:
Same comment as Nicole made on `show_twiss`. This looks like great command line functionality. Maybe we can move it somewhere else?

Reply:
moved to outputs.py